### PR TITLE
JWT Authentication Module

### PR DESCRIPTION
**Pull request by a-luz**
### Description
JWT module and interceptors for JWT authentication and content exchange.
Related issue: #191 

### Tester info
Besides running the unit tests for the jwt module (located in `test/xiana/jwt_test.clj`, there is also an example project in `examples/jwt` in which you can test the jwt interceptors

### CODE REVIEW

File: examples/jwt/src/backend/app/config.clj
Comment (g-krisztian) [inline examples/jwt/src/backend/app/config.clj:11]: I'd put this into config.edn

File: examples/jwt/src/backend/app/core.clj
Comment (g-krisztian) [inline examples/jwt/src/backend/app/core.clj:27]: this should be an jwt-init fn, and should be assoc the the xiana/jwt there

File: examples/jwt/src/frontend/controllers/views.cljs
Comment (g-krisztian) [inline examples/jwt/src/frontend/controllers/views.cljs:1]: the example has frontend, or it's just left here?

File: src/xiana/interceptor.clj
Comment (g-krisztian) [inline src/xiana/interceptor.clj:137]: I'd move these into a separated package, like xiana.jwt.interceptors

File: examples/jwt/src/backend/app/config.clj
Comment (g-krisztian) [inline examples/jwt/src/backend/app/config.clj:10]: These files should be configurable, and you can move the init function to the xiana too.

File: examples/jwt/src/backend/app/core.clj
Comment (g-krisztian) [inline examples/jwt/src/backend/app/core.clj:29]: the (config/config) merging its parameter if I'm right. It would be more elegant to pass the app-cfg to it

File: doc/jwt.md
Comment (g-krisztian) [inline doc/jwt.md:41]: Looks like this map isn't formatted well.

File: examples/jwt/src/backend/app/core.clj
Comment (g-krisztian) [inline examples/jwt/src/backend/app/core.clj:31]: Instead of merging `app-cfg` you can pass it to config/config.

File: src/xiana/jwt.clj
Comment (g-krisztian) [inline src/xiana/jwt.clj:15]:
```clojure
(assoc out-claims :exp (+ (util/now) exp-cfg)
                             :nbf (+ (util/now) nbf-cfg)
                             :iat (util/now))
```
Is a bit more elegant, and faster then creating a map, and mergin it.

**Review by g-krisztian — APPROVED** [review]

File: doc/jwt.md
Comment (soulflyer) [inline doc/jwt.md:56]: Probably "set" rather than "inform" and "will be used to sign"

File: doc/jwt.md
Comment (soulflyer) [inline doc/jwt.md:59]: " "

File: examples/jwt/README.md
Comment (soulflyer) [inline examples/jwt/README.md:24]: This gives me "Missing credentials"

Comment (g-krisztian) [reply examples/jwt/README.md:24]: @soulflyer You're right. 
@a-luz The `public-key` and the `private-key` variables are read from environment, right?

**Comment by g-krisztian** [comment]
Converted to draft, the initialization a bit messy, it must be cleaned up before merge.

**Review by a-luz — COMMENTED** [review]
I really liked the changes done here. I can't approve the PR as I've opened it, but it LGTM

**Review by soulflyer — APPROVED** [review]
LGTM !

**Comment by g-krisztian** [comment]
Just found some issues. 
 - The token's content should not be overwrite the session-data, but should be stored under another key is state.
 - @a-luz I don't get what the interceptor `:leave` part does. Is it encoding the body? It should work like this? I'm not familiar with the process.

**Comment by a-luz** [comment]
> Just found some issues.
> 
> * The token's content should not overwrite the session-data, but should be stored under another key is state.
> * @a-luz I don't get what the interceptor `:leave` part does. Is it encoding the body? It should work like this? I'm not familiar with the process.

Hey @g-krisztian, I had put the token's content to be the session-data  as it will hold the same kind of data as when using a normal login as it is being done now.

As for the `:leave` interceptor of the jwt-content, yes, that is what it should do. When using the jwt-content interceptor, the framework expects to always receive the body as a signed JWT and will always respond to the client as a signed JWT as well

**Comment by gmsvalente** [comment]
Keep in mind that we want to refresh tokens in the future

**Pull request merged**
